client/network/src/service/out_events.rs

```diff
@@ -0,0 +1,236 @@
+pub struct Sender {
+	inner: mpsc::UnboundedSender<Event>,
+	/// Clone of [`Receiver::stats`].
+	stats: Arc<Mutex<Option<Arc<Mutex<StatsCollect>>>>>,
+}
+
+impl fmt::Debug for Sender {
+	fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+		f.debug_tuple("Sender").finish()
+	}
+}
+
+/// Receiving side of a channel.
+pub struct Receiver {
+	inner: mpsc::UnboundedReceiver<Event>,
+	/// Initially contains `None`, and will be set to a value once the corresponding [`Sender`]
+	/// is assigned to an instance of [`OutChannels`].
+	stats: Arc<Mutex<Option<Arc<Mutex<StatsCollect>>>>>,
+}
+
+impl Stream for Receiver {
+	type Item = Event;
+
+	fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context) -> Poll<Option<Event>> {
+		if let Some(ev) = ready!(Pin::new(&mut self.inner).poll_next(cx)) {
+			let stats = self.stats.lock();
+			if let Some(stats) = stats.as_ref() {
+				decrease_stats(&mut *stats.lock(), &ev);
+			} else {
+				log::warn!("Inconsistency in out_events: event happened before sender associated");
+			}
+			Poll::Ready(Some(ev))
+		} else {
+			Poll::Ready(None)
+		}
+	}
+}
+
+impl fmt::Debug for Receiver {
+	fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+		f.debug_tuple("Receiver").finish()
+	}
+}
+
+impl Drop for Receiver {
+	fn drop(&mut self) {
+		// Empty the list to properly decrease the stats.
+		while let Some(Some(_)) = self.next().now_or_never() {}
+	}
+}
+
+/// Statistics about the content of the channel.
+#[non_exhaustive]
+pub struct StatsCollect {
+	/// For each protocol, total number of substream open notifications in the channel.
+	pub notifications_open_messages_count: HashMap<ConsensusEngineId, u64>,
+
+	/// For each protocol, total number of substream closed notifications in the channel.
+	pub notifications_closed_messages_count: HashMap<ConsensusEngineId, u64>,
+
+	/// For each protocol, the total size of all notifications stuck in the channel.
+	pub notifications_sizes_total: HashMap<ConsensusEngineId, u64>,
+
+	/// Number of DHT events in the channel.
+	///
+	/// > **Note**: DHT events are extremely rare at the moment, but this could be broken down
+	/// >			further if there's a need to.
+	pub dht_events: u64,
+}
+
+/// Collection of senders.
+pub struct OutChannels {
+	event_streams: Vec<Sender>,
+	/// The stats we collect. A clone of this is sent to each [`Receiver`] associated with this
+	/// object.
+	stats: Arc<Mutex<StatsCollect>>,
+}
+
+impl OutChannels {
+	/// Creates a new empty collection of senders.
+	pub fn new() -> Self {
+		OutChannels {
+			event_streams: Vec::new(),
+			stats: Arc::new(Mutex::new(StatsCollect {
+				notifications_open_messages_count: HashMap::new(),
+				notifications_closed_messages_count: HashMap::new(),
+				notifications_sizes_total: HashMap::new(),
+				dht_events: 0,
+			}))
+		}
+	}
+
+	/// Adds a new [`Sender`] to the collection.
+	pub fn push(&mut self, sender: Sender) {
```

Comment:
Why do we need  a `Sender` to ever live by its own? Why not have a `out_channels.new_channel()` function that adds the `Sender` to the set and returns the receiver?

Reply:
Ah, I see. Because the `NetworkService` needs to send it to the `NetworkWorker`.

Reply:
The API of `NetworkService::event_stream` creates a channel and instantly returns the receiver. At this point we don't have access to the `OutChannels`, which is stored in the `NetworkWorker`.

We could indeed send a message from the service to the worker to ask it to create a channel and return it, but that seems overly complicated.